### PR TITLE
Add publisher API test coverage

### PR DESCRIPTION
**Pull request by hidmic**
Depends on https://github.com/ros2/rmw/pull/252.

### CODE REVIEW

File: test_rmw_implementation/test/test_publisher.cpp [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+
+  const char * implementation_identifier = node->implementation_identifier;
+  node->implementation_identifier = "not-an-rmw-implementation-identifier";
+  pub = rmw_create_publisher(node, ts, topic_name, &rmw_qos_profile_default, &options);
+  node->implementation_identifier = implementation_identifier;
+  EXPECT_EQ(nullptr, pub);
+  rmw_reset_error();
+
+  pub = rmw_create_publisher(node, ts, nullptr, &rmw_qos_profile_default, &options);
+  ASSERT_EQ(nullptr, pub);
```
Comment (Blast545) [inline test_rmw_implementation/test/test_publisher.cpp:138]: IMO those tests are independent, and if one fails doesn't affect the others, so the ASSERT statements can be changed to EXPECT ones.
```suggestion
  EXPECT_EQ(nullptr, pub);
```

File: test_rmw_implementation/test/test_publisher.cpp [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, BasicTypes);
+  rmw_publisher_t * pub =
+    rmw_create_publisher(node, ts, topic_name, &rmw_qos_profile_system_default, &options);
+  ASSERT_NE(nullptr, pub) << rmw_get_error_string().str;
+  rmw_qos_profile_t qos_profile = rmw_qos_profile_unknown;
+  rmw_ret_t ret = rmw_publisher_get_actual_qos(pub, &qos_profile);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+  EXPECT_NE(rmw_qos_profile_system_default.history, qos_profile.history);
+  EXPECT_NE(rmw_qos_profile_unknown.history, qos_profile.history);
+  EXPECT_NE(rmw_qos_profile_system_default.reliability, qos_profile.reliability);
```
Comment (Blast545) [inline test_rmw_implementation/test/test_publisher.cpp:214]: The publisher is created with `rmw_qos_profile_system_default` then why when reading the `qos_profile` do not match to this? It seems odd, mind asking a comment on this?

Comment (Blast545) [reply test_rmw_implementation/test/test_publisher.cpp:214]: I can't see the difference between this section vs the last test case `get_actual_qos`

Comment (hidmic) [reply test_rmw_implementation/test/test_publisher.cpp:214]: > The publisher is created with rmw_qos_profile_system_default then why when reading the qos_profile do not match to this? It seems odd, mind asking a comment on this?

`_SYSTEM_DEFAULT` are sentinel values, expect to be replaced by concrete policies. This test verifies that's the case in a generic way (i.e. can't test for a specific policy as it is implementation dependent).

> I can't see the difference between this section vs the last test case get_actual_qos

That other test uses concrete policies and so it expects them to be honored.

Just as an FYI, liveliness, lifespan and deadline QoS policies are less consistent in this regard and thus why they are currently not verified. I plan to come back to them at a later point in time.

File: test_rmw_implementation/test/test_publisher.cpp (hunk 1 above)
Comment (hidmic) [reply test_rmw_implementation/test/test_publisher.cpp:138]: Good point. Also, there's a test above that should be two separate tests.

File: test_rmw_implementation/test/test_publisher.cpp [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+  {
+    init_options = rmw_get_zero_initialized_init_options();
+    rmw_ret_t ret = rmw_init_options_init(&init_options, rcutils_get_default_allocator());
+    ASSERT_EQ(RMW_RET_OK, ret) << rcutils_get_error_string().str;
+    init_options.enclave = rcutils_strdup("/", rcutils_get_default_allocator());
+    ASSERT_STREQ("/", init_options.enclave);
+    context = rmw_get_zero_initialized_context();
+    ret = rmw_init(&init_options, &context);
+    ASSERT_EQ(RMW_RET_OK, ret) << rcutils_get_error_string().str;
+    const char * const node_name = "my_test_node";
```
Comment (brawner) [inline test_rmw_implementation/test/test_publisher.cpp:45]: Why not just constexpr? Here and below for string literals.

File: test_rmw_implementation/test/test_publisher.cpp [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    ret = rmw_init_options_fini(&init_options);
+    EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+  }
+
+  rmw_init_options_t init_options;
+  rmw_context_t context;
+  rmw_node_t * node;
+};
+
+class CLASSNAME (TestPublisherUse, RMW_IMPLEMENTATION) :
```
Comment (brawner) [inline test_rmw_implementation/test/test_publisher.cpp:68]: Typically it's common to group classes with the tests yah? Should this class definition be moved to line 225?

File: test_rmw_implementation/test/test_publisher.cpp [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+  pub = rmw_create_publisher(node, ts, topic_name, &rmw_qos_profile_default, &options);
+  node->implementation_identifier = implementation_identifier;
+  EXPECT_EQ(nullptr, pub);
+  rmw_reset_error();
+
+  pub = rmw_create_publisher(node, ts, nullptr, &rmw_qos_profile_default, &options);
+  ASSERT_EQ(nullptr, pub);
+  rmw_reset_error();
+
+  const char * const topic_name_with_spaces = "foo bar";
```
Comment (brawner) [inline test_rmw_implementation/test/test_publisher.cpp:141]: If relative topic names are also invalid, shouldn't this start with a leading '/'?

File: test_rmw_implementation/test/test_publisher.cpp (hunk 2 above)
Comment (brawner) [reply test_rmw_implementation/test/test_publisher.cpp:214]: This test probably deserves a comment with this explanation.

File: test_rmw_implementation/test/test_publisher.cpp (hunk 5 above)
Comment (hidmic) [reply test_rmw_implementation/test/test_publisher.cpp:141]: Good point.

File: test_rmw_implementation/test/test_publisher.cpp (hunk 4 above)
Comment (hidmic) [reply test_rmw_implementation/test/test_publisher.cpp:68]: Sounds reasonable, moved.

File: test_rmw_implementation/test/test_publisher.cpp (hunk 3 above)
Comment (hidmic) [reply test_rmw_implementation/test/test_publisher.cpp:45]: Sounds good. I expect the compiler to inline references to the global literal anyways.

File: test_rmw_implementation/test/test_publisher.cpp [hunk 6]
```diff
@@ -42,8 +42,8 @@ class CLASSNAME (TestPublisher, RMW_IMPLEMENTATION) : public ::testing::Test
     context = rmw_get_zero_initialized_context();
     ret = rmw_init(&init_options, &context);
     ASSERT_EQ(RMW_RET_OK, ret) << rcutils_get_error_string().str;
-    const char * const node_name = "my_test_node";
-    const char * const node_namespace = "/my_test_ns";
+    constexpr const char * node_name = "my_test_node";
```
Comment (brawner) [inline test_rmw_implementation/test/test_publisher.cpp:45]:
```suggestion
    constexpr char node_name[] = "my_test_node";
```

Comment (hidmic) [reply test_rmw_implementation/test/test_publisher.cpp:45]: Sure, yeah. But for the record, that may force a stack allocation we don't need.

Comment (hidmic) [reply test_rmw_implementation/test/test_publisher.cpp]: For posterity, it turns out that, in exchange, using `constexpr` precludes string literals entirely and reduces somewhat the size of the binary. Thanks for the tip @brawner !

**Comment by hidmic** [comment]
CI up `test_rmw_implementation` against all RMW implementations:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11630/)

**Comment by brawner** [comment]
Why does the rpr build fail for connextdds? [Build Status](http://build.ros2.org/job/Rpr__rmw_implementation__ubuntu_focal_amd64/45/)

**Comment by hidmic** [comment]
@brawner I have to merge related RMW PRs and release those repos to `rolling` before the PR job can pass.

**Comment by hidmic** [comment]
@ros-pull-request-builder retest this please.

**Comment by hidmic** [comment]
Alright, all green! Going in.

**Pull request merged**